container.go

```diff
@@ -109,96 +99,145 @@ func (c Container) invoke(function interface{}) (interface{}, error) {
 	return values[0].Interface(), nil
 }
 
+// make resolves the binding and returns the concrete.
+// Search up any parent container scopes if the binding is not found in current scope.
+func (c *Container) make(ctx context.Context, t reflect.Type, name string) (interface{}, error) {
+	current := c
+	var binding *binding
+
+	for {
+		if found, exist := current.bindings[t][name]; exist {
+			binding = found
+			break
+		}
+
+		if current.parent == nil {
+			break
+		} else {
+			current = current.parent
+		}
+	}
+
+	if binding == nil {
+		return nil, errors.New("container: no binding found for: " + t.String())
+	}
+
+	return binding.make(ctx, current)
+}
+
 // arguments returns the list of resolved arguments for a function.
-func (c Container) arguments(function interface{}) ([]reflect.Value, error) {
+func (c *Container) arguments(ctx context.Context, function interface{}) ([]reflect.Value, error) {
 	reflectedFunction := reflect.TypeOf(function)
 	argumentsCount := reflectedFunction.NumIn()
 	arguments := make([]reflect.Value, argumentsCount)
+	contextType := reflect.TypeOf((*context.Context)(nil)).Elem()
 
 	for i := 0; i < argumentsCount; i++ {
 		abstraction := reflectedFunction.In(i)
-		if concrete, exist := c[abstraction][""]; exist {
-			instance, err := concrete.make(c)
-			if err != nil {
-				return nil, err
+
+		if abstraction.Implements(contextType) {
```

Comment:
I'm still not sure this is exactly what we want - but I admit that this is a place the intersection of the injection of params and how "context" is sort of a special param in the go ecosystem.

My general thought was we'd have a design like this:

1. Instead of the `WithContext` and regular versions, just have a single version that takes a `Context`.  Folks can pass in `context.Background()` or `context.Todo()`.  If we want to keep the `WithContext` versions (for API compat reasons with merging back with upstream at some point, I think that's fine), but I think the non context versions should be documented as just calling the `WithContext` version with `context.Background()`
2. When we find the resolver for the type, we look at the first argument.  If it is a context, assign it the context value that was passed in to `arguments`.  Otherwise (and for every other argument) we just do what we would have done before (resolve the concrete from the container).

This does mean that if you register a `context.Context` in the container and run a resolver that looks like `func(ctx context.Context, otherCtx context.Context)` you'd end up with the two different contexts being passed into the resolver (which is sort of confusing, I grant, but I also think that in general folks will not be registering a context in the container, based on how folks use `Context` in go)

To me this feel the most natural for folks writing resolvers - you can take a context as the first argument and the value gets the `context` that's being used for the resolve operation.